diff --git a/libcli/smbreadline/smbreadline.c b/libcli/smbreadline/smbreadline.c
index 6929209be20..012d08e04bc 100644
--- a/libcli/smbreadline/smbreadline.c
+++ b/libcli/smbreadline/smbreadline.c
@@ -146,8 +146,12 @@ char *smb_readline(const char *prompt, void (*callback)(void),
 	}
 
 #ifdef HAVE_DECL_RL_EVENT_HOOK
-	if (callback)
-		rl_event_hook = (rl_hook_func_t *)callback;
+  if (callback)
+#if defined(_RL_FUNCTION_TYPEDEF)
+    rl_event_hook = (rl_hook_func_t *)callback;
+#else
+  rl_event_hook = (Function *)callback;
+#endif
 #endif
 	ret = readline(prompt);
 	if (ret && *ret)
